### PR TITLE
[MachO] Detect overflow in section offset.

### PR DESCRIPTION
**Pull request by efriedma-quic**
The section offset field is only 32 bits; if the computed section offset is larger, make sure we don't emit a corrupt object file.

### CODE REVIEW

File: llvm/lib/MC/MachObjectWriter.cpp [hunk 1]
```diff
@@ -277,6 +277,8 @@ void MachObjectWriter::writeSection(const MCAssembler &Asm,
     W.write<uint32_t>(VMAddr);      // address
     W.write<uint32_t>(SectionSize); // size
   }
+  if (!isUInt<32>(FileOffset))
+    report_fatal_error("Cannot encode offset of section");
   W.write<uint32_t>(FileOffset);
 
   W.write<uint32_t>(Log2(Section.getAlign()));
```
Comment (jroelofs) [inline llvm/lib/MC/MachObjectWriter.cpp:284]: Mind adding another one for `RelocationStart`? Looks like it might have the same problem.

**Review by jroelofs — APPROVED** [review]

File: llvm/lib/MC/MachObjectWriter.cpp (hunk 1 above)
Comment (MaskRay) [reply llvm/lib/MC/MachObjectWriter.cpp:284]: We can call `Asm.getContext().reportError` instead to not trigger a crash.

Comment (MaskRay) [reply llvm/lib/MC/MachObjectWriter.cpp:284]: Perhaps let the caller (`MachObjectWriter::writeObject`) check this and early return.

Comment (efriedma-quic) [reply llvm/lib/MC/MachObjectWriter.cpp:284]: The point of the early return would be to avoid duplicate error messages?  That makes sense, I guess.

Comment (efriedma-quic) [reply llvm/lib/MC/MachObjectWriter.cpp:283]: Tried rearranging as suggested.

**Review by jroelofs — APPROVED** [review]

File: llvm/lib/MC/MachObjectWriter.cpp [hunk 2]
```diff
@@ -904,6 +908,18 @@ uint64_t MachObjectWriter::writeObject(MCAssembler &Asm) {
     unsigned Flags = Sec.getTypeAndAttributes();
     if (Sec.hasInstructions())
       Flags |= MachO::S_ATTR_SOME_INSTRUCTIONS;
+    if (!cast<MCSectionMachO>(Sec).isVirtualSection() &&
+        !isUInt<32>(SectionStart)) {
+      Asm.getContext().reportError(
+          SMLoc(), "Cannot encode offset of section; object file too large");
```
Comment (MaskRay) [inline llvm/lib/MC/MachObjectWriter.cpp:914]: `reportError` messages are generally not capitalized per https://llvm.org/docs/CodingStandards.html#error-and-warning-messages

File: llvm/test/MC/MachO/section-offset-overflow.s [hunk 3]
```diff
@@ -0,0 +1,9 @@
+// RUN: not llvm-mc -triple x86_64-apple-macosx -filetype=obj -o /dev/null %s 2>&1 | FileCheck  %s
+
+// CHECK: Cannot encode offset of section
```
Comment (MaskRay) [inline llvm/test/MC/MachO/section-offset-overflow.s:3]: check the `error:` prefix of the diagnostic

**Review by MaskRay — APPROVED** [review]

**Pull request merged**